File: 2023/05/21/min-cost-to-connect-all-points.py


class Union:
    def __init__(self, size:int):
        self.data = [i for i in range(size)]

    def rep(self, x: int) -> int:
        stack = [x]
        while stack:
            current = stack[-1]
            if self.data[current] == current:
                # for element in stack:
                #     self.data[element] = current
                return current
            else:
                stack.append(self.data[current])


    def join(self, x: int, y: int):
        x_rep = self.rep(x)
        y_rep = self.rep(y)
        
        self.data[y_rep] = x_rep


class Solution:
    def manDistance(self, p1, p2):
        x = abs(p1[0] - p2[0])
        y = abs(p1[1] - p2[1])
        return x + y

    def minCostConnectPoints(self, points):
        distances = []
        union = Union(len(points))
        for x in range(len(points)):
            for y in range(len(points)):
                distances.append([x, y, self.manDistance(points[x], points[y])])

        distances.sort(key=lambda a: a[2])
        minCost = 0
        counter = 0

        for distance in distances:
            if union.rep(distance[0]) != union.rep(distance[1]):
                union.join(distance[0], distance[1])
                minCost += distance[2]
                counter += 1
                if counter == len(points) - 1:
                    break

        return minCost

File: 2023/05/21/min-cost-to-connect-all-points.dart
class Union {
  late List<int> data;

  Union(int size) {
    this.data = List.generate(size, (index) => index);
  }

  int rep(int x) {
    List<int> stack = [x];
    while (stack.isNotEmpty) {
      int current = stack.last;
      if (current == data[current]) {
        return current;
      } else {
        stack.add(this.data[current]);
      }
    }
    throw Exception();
  }

  void join(int x, int y) {
    int x_rep = rep(x);
    int y_rep = rep(y);

    data[x_rep] = data[y_rep];
  }
}

class Solution {
  int manDistance(List<int> p1, List<int> p2) {
    int x = (p1[0] - p2[0]).abs();
    int y = (p1[1] - p2[1]).abs();
    return x + y;
  }

  int minCostConnectPoints(List<List<int>> points) {
    List<List<int>> distances = [];
    Union union = Union(points.length);
    for (var x = 0; x < points.length; x++) {
      for (var y = 0; y < points.length; y++) {
        distances.add([x, y, manDistance(points[x], points[y])]);
      }
    }
    distances.sort((a,b) =>a[2].compareTo(b[2]));
    int minCost = 0;
    int counter = 0;

    for (var distance in distances) {
      if (union.rep(distance[0]) != union.rep(distance[1])) {
        union.join(distance[0], distance[1]);
        minCost += distance[2];
        counter += 1;
        if (counter == points.length - 1) {
          break;
        }
      }
    }
    return minCost;
  }
}
